### PR TITLE
Fix torch.optim.Adagrad for CUDA

### PR DESCRIPTION
**Pull request by Roffild**
Fix this:
```python
import torch
model = torch.nn.Sequential(torch.nn.Linear(1, 1))
optim = torch.optim.Adagrad(model.parameters())
loss = torch.nn.MSELoss()
inp = torch.tensor([[45.0]], device="cuda")
outp = torch.tensor([[99.0]], device="cuda")

model.to(device="cuda")
optim.zero_grad()
loss(model(inp), outp).backward()
optim.step()
```
```
  File "torch\optim\functional.py", line 48, in adagrad
    state_sum.addcmul_(grad, grad, value=1)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

### CODE REVIEW

File: torch/optim/adagrad.py [hunk 1]
```diff
@@ -38,18 +38,6 @@ def __init__(self, params, lr=1e-2, lr_decay=0, weight_decay=0, initial_accumula
                         initial_accumulator_value=initial_accumulator_value)
         super(Adagrad, self).__init__(params, defaults)
 
-        for group in self.param_groups:
-            for p in group['params']:
-                state = self.state[p]
-                state['step'] = 0
-                state['sum'] = torch.full_like(p, initial_accumulator_value, memory_format=torch.preserve_format)
-
-    def share_memory(self):
```
Comment (wanchaol) [inline torch/optim/adagrad.py:47]: `shared_memory` is still useful, we shouldn't delete this.

**Review by wanchaol — CHANGES_REQUESTED** [review]
Just looked around all other optimizers, it seems like only Adagrad not doing the lazy initialization, so fixing this alone should be fine. Please revert the removal of `share_memory`, fix the CI errors (possibly related to this [functional adagrad](https://github.com/pytorch/pytorch/blob/master/torch/distributed/optim/functional_adagrad.py#L51) ), and it should be good to go.

**Comment by Roffild** [comment]
Only in step() is p.grad located on the correct device. Therefore, assigning values only in step().

Unlike other optimizers, everything is more confusing here.

I found a bug and fixed it as in other optimizers, but need a more experienced programmer in this code.

**Comment by ngimel** [comment]
Yeah, the bug fix is fine, but please don't delete `share_memory` function

**Comment by Roffild** [comment]
I have pasted share_memory.
But I doubt the need for this function, because the variable initialization is now in a different place.

**Comment by wanchaol** [comment]
> I have pasted share_memory.
> But I doubt the need for this function, because the variable initialization is now in a different place.

I looked into this again and I don't know if this change is legit or not. In particular, can you describe your use case about why you need to do `model.to(cuda)` after the optimizer construction? 

Usually you can just construct the optimizer after you move the model to the desired device. Deleting `share_memory` means a API change, and even if not deleting it, it seems that user couldn't calls `share_memory` before the first `step()` after this change, which is BC change. People use `share_memory` quite often (i.e. in multi-process hogwild), so we should be careful on this. 

I know that there're cases where we do lazy construction of optimizer states for other optimizers, but I don't think they are there for the use case stated in the description. cc @albanD do you think we should lazy initialize Adagrad states?

**Comment by ngimel** [comment]
That's right, the behavior wasn't changed by a recent switch to functional optimizers, the states were always initialized on construction https://github.com/pytorch/pytorch/blame/8ab2ad306d247679fc1ab2b5941f9e978f8e178b/torch/optim/adagrad.py#L44

**Comment by Roffild** [comment]
> In particular, can you describe your use case about why you need to do model.to(cuda) after the optimizer construction?

[self.state = {}](https://github.com/pytorch/pytorch/blob/v1.8.0/torch/optim/optimizer.py#L45) for all optimizers.

`model.to()` is not part of the initialization and can be called in any order.

Only for `model.to(device); optim.step()` is guaranteed to be the correct device. Therefore, the initialization in `step()`:
https://github.com/pytorch/pytorch/blob/v1.8.0/torch/optim/adadelta.py#L69
https://github.com/pytorch/pytorch/blob/v1.8.0/torch/optim/adam.py#L86
https://github.com/pytorch/pytorch/blob/v1.8.0/torch/optim/adamax.py#L62
https://github.com/pytorch/pytorch/blob/v1.8.0/torch/optim/asgd.py#L58
https://github.com/pytorch/pytorch/blob/v1.8.0/torch/optim/rmsprop.py#L88
...

Only in Adagrad this bug.

And `share_memory` is missing from optimizers. I made Adagrad according to Pytorch standards. If my fix breaks something else, then there is also a bug.

**Comment by albanD** [comment]
Hi,

The failing test is actually a test for LazyModules that check that they work well with optimizers.
The failure is the special case for Adagrad because it was not lazy and so was failing.
Now that you fixed Adagrad, you should update that test and remove the special case. Adagrad should now behavior just like the other optimizers.

I am not sure why we have a share_memory on this optimizer though.
If we fix this initialization, then I would agree that it is safer to make that method throw an error if the buffers are not initialized (as otherwise people that used to call it and get the state dict before using the optimizer will silently get an empty state dict).
Does that sound good?

Also @Roffild could you rebase on top of master to remove the merge conflicts and make sure that all other CI builds pass.

**Comment by Roffild** [comment]
The test has fixed.

[There are no merge conflicts.](https://github.com/pytorch/pytorch/compare/master...Roffild:patch-2)

**Review by ngimel — APPROVED** [review]

**Comment by Roffild** [comment]
Why isn't this patch merged?

**Comment by ngimel** [comment]
Because it is breaking a lot of internal tests that we didn't have time to resolve.

**Comment by Roffild** [comment]
The patch is ready. A year has passed. Maybe it's time to merge?

**Comment by Roffild** [comment]
Oh, no, no, no... This patch is alive!

**Pull request closed without merging**